Consider future deletionTimestamps when validating bound tokens

liggitt opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When an object bound to a serviceaccount token is deleted with a grace period, the bound token should continue to be valid until the grace period has expired. This allows a pod that makes API requests while handling a sigterm in response to a pod deletion to continue to be able to authenticate for the duration of the grace period.

Part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190806-serviceaccount-tokens.md

**Does this PR introduce a user-facing change?**:
```release-note
Service account tokens bound to pods can now be used during the pod deletion grace period.
```

/sig auth
/cc @mikedanese

liggitt committed:
Consider future deletionTimestamps when validating bound tokens

liggitt commented:
/priority important-soon

mikedanese commented on pkg/serviceaccount/claims.go:
```diff
@@ -120,7 +121,7 @@ func (v *validator) Validate(_ string, public *jwt.Claims, privateObj interface{
 		return nil, err
 	}
-	if serviceAccount.DeletionTimestamp != nil {
+	if serviceAccount.DeletionTimestamp != nil && !serviceAccount.DeletionTimestamp.Time.After(nowTime) {
```
What about the grace period? If a pod is gracefully shutting down, should it still be able to authenticate?

liggitt replied:
the deletionTimestamp reflects the grace period... if I delete at time T with a grace period of 30 seconds, the deletionTimestamp is set to T+30

mikedanese commented on pkg/serviceaccount/claims_test.go:
```diff
@@ -182,3 +185,207 @@ func TestClaims(t *testing.T) {
+			v := &validator{tc.getter}
+			_, err := v.Validate("", tc.claims, tc.private)
+			if err != nil {
+				if tc.expectErr {
```
nit
```
if err != nil && !tc.expectErr {
  t.Fatal(err)
}
```
has better symmetry with the next check.

liggitt replied:
updated

mikedanese commented:
We give one minute of leeway for nbf, exp claims:

https://github.com/kubernetes/kubernetes/blob/3842a92f5ffa7ceebf2058700c2494906f977800/vendor/gopkg.in/square/go-jose.v2/jwt/validation.go#L23-L24

Should we do the same here?

liggitt commented:
re: leeway, probably so. once the object is actually deleted, the validation fails immediately, so we still get the benefit of immediate revocation once the kubelet terminates the process and deletes the pod.

updated, added testcases and cleaned them up to be more readable

mikedanese commented:
/lgtm
/retest
